Sprint 21

ArtemDubovitskiy opened the pull request:
Прошу проверить самостоятельную задачу по 21 спринту:

Связаны между собой сетевой слой и графический интрефейс приложения

Спасибо!)

ArtemDubovitskiy committed:
Внесены изменения:

- экран настроек переписан на MVVM
- константы вынесены в отдельный enum (файл)
- состояния вью вынесены в отдельный файл
-

ArtemDubovitskiy committed:
Добавлено:

- экран польз. соглашения переписан на MVVM
- исправлено состояние переключателя на экране настроек при перезапуске приложения

ArtemDubovitskiy committed:
Экран Stories переписан на MVVM

ArtemDubovitskiy committed:
Добавлен CopyrightRequestService, переделан экран настроек

ArtemDubovitskiy committed:
Исправлено:

- отображение progressView только в месте загрузки данных (текста)
- цикличная загрузка
- при ошибке загрузки данных моковый текст

ArtemDubovitskiy committed:
Merge pull request #4 from ArtemDubovitskiy/sprint_21_test

Sprint 21 test

ArtemDubovitskiy committed:
Добавлено:

- StationsListRequestService (загрузка всех станций из API)
- StationsListService (формат станций в модель)
- поправлена viewModel (подключен сервис)
- исправлены экраны выбора города и станций (добавлен scroll)
- добавлен ProgressView + состояние на время загрузки списка станций

ArtemDubovitskiy committed:
Добавлено:

- SearchRequestService (загрузка маршрутов из API)
- SearchService (формат маршрутов в модель)
- поправлена viewModel (подключен сервис)

ArtemDubovitskiy committed:
Рефакторинг:

- добавлена CitiesViewModel
- исправлена основная модель (MainSearchViewModel)
- временно удалены состояния с вью без моделей
- отредактирован CitiesScreenView (navigationBar)

ArtemDubovitskiy committed:
Рефакторинг:

- добавлена ScheduleViewModel
- из MainSearchViewModel удалены все лишние методы
- в ScheduleScreenView добавлена загрузка рейсов
- поправлена StationsScreenView (убрал состояния, т.к. нет работы с сетью)

ArtemDubovitskiy committed:
Исправлено:

- загрузка лого перевозчика в расписании и карточке перевозчика
- добавлены заглушки, если изображения нет

ArtemDubovitskiy committed:
Исправления:

- формат времени отправления и времени прибытия в расписании (часы / мин)
- фильтрация расписания по дате отправления (не по времени)
- время рейса локализовано

ArtemDubovitskiy committed:
Рефакторинг:

- удален закомментированный код, лишние сущности

ArtemDubovitskiy committed:
исправления

ArtemDubovitskiy committed:
добавлена загрузка изображения на экране инфо о перевозчике

So1ovyov commented on TravelSchedule/Constants/Constants.swift:
```diff
@@ -0,0 +1,12 @@
+import Foundation
+
+enum Constants {
+    static let apiKey = " "
```
Почему `apiKey` пустой, это не настолько важная информация, которой можно воспользоваться в своих корыстных целях, так что можно вернуть. Если ты забыл добавить, то давай сделаем это.

ArtemDubovitskiy replied:
Исправил, ранее как то были замечания касательно ключей (перестал их добавлять)

So1ovyov replied:
@ArtemDubovitskiy странно, что за это пишут какие-то замечания...

So1ovyov commented on TravelSchedule/Extensions/DateComponentsFormatter+Ext.swift:
```diff
@@ -0,0 +1,20 @@
+//
+import Foundation
+
+extension DateComponentsFormatter {
```
На самом деле расширения намного тяжеловеснее в плане производительность, чем какие-то отдельные сущности с логикой. Поэтому предлагаю расширения для `DateComponentsFormatter` и `DateFormatter` вынести в отдельную сущность `Helper`
```
enum DateFormatterHelper {

    static let formatter: DateComponentsFormatter = {
        var calendar = Calendar.current
        calendar.locale = Locale(identifier: "ru")
        let formatter = DateComponentsFormatter()
        formatter.calendar = calendar
        // Можно поменять на часы и полное наименование (некоторые рейсы актуальны в минутах)
        formatter.allowedUnits = [.hour, .minute]
        formatter.unitsStyle = .abbreviated
        return formatter
    }()

    static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.locale = Locale(identifier: "ru_Ru")
        return formatter
    }()
}
```

So1ovyov commented on TravelSchedule/Model/Carrier.swift:
```diff
@@ -9,8 +9,8 @@ import Foundation
 struct Carrier: Hashable, Identifiable {
     let id = UUID()
     let title: String
+    let logo_svg: String
```
Наименование констант, переменных и методов должно быть в стиле CamelCase и начинаться с нижнего регистра.
В данном случае необходимо создать `enum CodingKeys`, а нейминг исправить `logo_svg` -> `logoSvg`

ArtemDubovitskiy replied:
Спасибо, упустил)

So1ovyov commented on TravelSchedule/Model/MockData.swift:
Нужно будет указать где-нибудь `TODO`, что моки остались для `Preview` и никак не используются иначе.

ArtemDubovitskiy replied:
+

So1ovyov commented on TravelSchedule/Model/Station.swift:
```diff
@@ -9,4 +9,5 @@ import Foundation
 struct Station: Hashable, Identifiable {
```
Структуры, которые используются в сервисах и их ответах, нужно помечать как `Sendable`

ArtemDubovitskiy replied:
Спасибо!

So1ovyov commented on TravelSchedule/NetworkRequestService/CopyrightRequestService.swift:
```diff
@@ -0,0 +1,35 @@
+    func getCopyright() async throws -> CopyrightSchedule
+}
+
+actor CopyrightRequestService: CopyrightRequestServiceProtocol, @unchecked Sendable {
```
Почему используется `@unchecked Sendable` ? Стандартный `Sendable` также корректно отрабатывает без ошибок
Предлагаю во всех сущностях избавиться от `@unchecked`

ArtemDubovitskiy replied:
+

So1ovyov commented on TravelSchedule/NetworkRequestService/CopyrightService.swift:
```diff
@@ -0,0 +1,29 @@
+import Foundation
+import OpenAPIURLSession
+
+final class CopyrightService {
```
Также по ТЗ все сервисы должны быть `actor`
```
actor CopyrightService {
```

ArtemDubovitskiy replied:
+

So1ovyov commented on TravelSchedule/NetworkRequestService/SearchService.swift:
```diff
@@ -0,0 +1,75 @@
+import Foundation
+import OpenAPIURLSession
+
+final class SearchService {
```
Давай использовать `actor`

ArtemDubovitskiy replied:
+

So1ovyov commented on TravelSchedule/NetworkRequestService/StationsListService.swift:
```diff
@@ -0,0 +1,60 @@
+import Foundation
+import OpenAPIURLSession
+
+final class StationsListService {
```
Если забыть про `actor`, то `xcode` подскажет что нужно исправить:
```
actor StationsListService {
```
Иначе
![CleanShot 2024-07-23 at 14 21 14@2x](https://github.com/user-attachments/assets/3c842d21-d315-4505-8fb9-fb9a5ce15a32)

So1ovyov commented on TravelSchedule/View/CitiesScreen/CitiesScreenView.swift:
```diff
@@ -8,12 +8,11 @@
 struct CitiesScreenView: View {
-    
+    @EnvironmentObject var mainSearchViewModel: MainSearchViewModel
+    @StateObject var viewModel = CitiesViewModel()
```
Давай для `ViewModel` использовать типа в котором он объявлен `ObservableObject` -> `@ObservedObject`
```
@ObservedObject var viewModel = CitiesViewModel()
```

ArtemDubovitskiy replied:
Все модели поправил на `@ObservableObject` -> `@ObservedObject`
Но пока не могу решить вопрос с состояниями вью при их изменениями.

So1ovyov replied:
@ArtemDubovitskiy по коду кажется, что все должно работать корректно.

ArtemDubovitskiy replied:
@So1ovyov спасибо за ревью!

So1ovyov commented on TravelSchedule/View/ScheduleScreen/CarrierInfoScreen/CarrierInfoScreenView.swift:
```diff
@@ -11,91 +11,89 @@ struct CarrierInfoScreenView: View {
+    @Environment(\.dismiss) private var dismiss
     @Environment(\.openURL) private var openURL
-    @EnvironmentObject var viewModel: ScheduleViewModel
+    @EnvironmentObject var viewModel: MainSearchViewModel
```
Проверь все вью, в которых используешь вью модели и примени к ним `property wrapper` типа  `@ObservedObject`
```
@ObservedObject var viewModel = MainSearchViewModel()
```

ArtemDubovitskiy replied:
Первоначально у меня была 1 вью модель на все представления.
Далее в процессе рефакторинга я оставил Main для пробрасывания значений из этой модели для всех представлений. но почитав по более я понимаю проблема в том что можно допустить утечки памяти если неаккуратно использовать)
Спасибо!

So1ovyov commented on TravelSchedule/View/ScheduleScreen/ScheduleScreenView.swift:
```diff
@@ -9,21 +9,27 @@ import SwiftUI
-    @EnvironmentObject var viewModel: ScheduleViewModel
-    @Environment(\.dismiss) private var dismiss // заглушка
+    @EnvironmentObject var mainSearchViewModel: MainSearchViewModel
+    @StateObject var viewModel = ScheduleViewModel()
```
Не пропусти пожалуйста, здесь вообще `StateObject`, который не нужно использовать
```
@ObservedObject var viewModel = ScheduleViewModel()
```

ArtemDubovitskiy replied:
+

So1ovyov commented on TravelSchedule/View/SettingsScreen/WebView.swift:
```diff
@@ -21,12 +21,6 @@ struct WebView: UIViewRepresentable {
-         Но кастомный CSS не применяется к URL: https://yandex.ru/legal/practicum_offer (стр. 30-46)
-         Для проверки темной темы добавлен URL: https://developer.apple.com в SettingsScreenView
-         Прошу не считать критическим замечанием.
-         */
```
Все выглядит отлично, я бы не сказал, что здесь может быть какой-то крит

So1ovyov commented on TravelSchedule/View/StartScreen/MainSearchView.swift:
```diff
@@ -9,7 +9,7 @@ import SwiftUI
 struct MainSearchView: View {
     @State private var path: [Destination] = []
-    @EnvironmentObject var viewModel: ScheduleViewModel
+    @EnvironmentObject var mainSearchViewModel: MainSearchViewModel
```
`@ObservedObject var mainSearchViewModel = MainSearchViewModel()`

So1ovyov commented on TravelSchedule/View/StartScreen/MainSearchView.swift:
```diff
@@ -9,7 +9,7 @@ import SwiftUI
     @State private var path: [Destination] = []
-    @EnvironmentObject var viewModel: ScheduleViewModel
+    @EnvironmentObject var mainSearchViewModel: MainSearchViewModel
     // TODO: Добавить локализацию
```
Очень хорошая привычка оставлять `TODO` как напоминание себе и информация для ревьюера, молодец!

So1ovyov commented on TravelSchedule/ViewModel/CitiesViewModel.swift:
```diff
@@ -0,0 +1,32 @@
+    private let stationService = StationsListService()
+    
+    // MARK: - Public Methods
+    func getCities() async {
```
А почему именно через отдельное свойство `cities` ? Разве нельзя через метод возвращать массив или ошибку ? 
Например что-то подобное ?
```
func getCities() async throws -> [City] {
```

So1ovyov replied:
Главное структуру подписать под `Sendable`
```
struct City: Hashable, Identifiable, Sendable {
```

So1ovyov commented on TravelSchedule/ViewModel/SettingsViewModel.swift:
```diff
@@ -0,0 +1,37 @@
+final class SettingsViewModel: ObservableObject, @unchecked Sendable {
+    @AppStorage("isDarkScheme") private var isDarkScheme: Bool = false
+    @Published var isDarkSchemeOn: Bool = false
+    @Published var copyright_text: String = ""
```
Наименование в `CamelCase`
```
@Published var copyrightText: String = ""
```

ArtemDubovitskiy replied:
+

ArtemDubovitskiy committed:
Исправления по замечаниям ревью:

- добавлен apiKey
- переименованы константы
- структуры помечены как Sendable
- все сервисы - actor
- для всех viewModels использован тип @ObservedObject

ArtemDubovitskiy committed:
Исправления:

- все вьюмодели ObservedObject
- удалена дата из запроса расписания (т.к. возвращается массив только на запрашиваемый день)

ArtemDubovitskiy committed:
Исправления:

- фильтрация городов по наличию ЖД станций
- оставлен только список ЖД станций (для выбора только поездов)
- удалена дата в запросе расписания (для выдачи рейсов на текущую дату и позже)
- добавлен enum DateFormatterHelper для форматирования дат
